File: mobile/lib/core/protocol/protocol_metrics.dart
import 'dart:typed_data';

/// Port of `in.meshsetu.protocol` metrics/debug helpers
/// (Kotlin `ProtocolMetrics.kt`).
class ProtocolMetric {
  const ProtocolMetric({
    required this.timeMs,
    this.eventId,
    this.peer,
    required this.kind,
    this.value,
    this.detail,
  });

  final int timeMs;
  final String? eventId;
  final String? peer;
  final String kind;
  final int? value;
  final String? detail;
}

class JsonLineMetricSink {
  JsonLineMetricSink(this._sink);

  final StringSink _sink;

  static String _quote(String value) =>
      '"${value.replaceAll('\\', '\\\\').replaceAll('"', '\\"').replaceAll('\n', '\\n')}"';

  void write(ProtocolMetric metric) {
    final fields = <String>[
      '"time_ms":${metric.timeMs}',
      '"kind":${_quote(metric.kind)}',
    ];
    if (metric.eventId != null) {
      fields.add('"event_id":${_quote(metric.eventId!)}');
    }
    if (metric.peer != null) {
      final truncated = metric.peer!.length > 12
          ? metric.peer!.substring(0, 12)
          : metric.peer!;
      fields.add('"peer":${_quote(truncated)}');
    }
    if (metric.value != null) fields.add('"value":${metric.value}');
    if (metric.detail != null) fields.add('"detail":${_quote(metric.detail!)}');
    _sink.write('{${fields.join(',')}}\n');
  }
}

class LossyFrameInterceptor {
  LossyFrameInterceptor({this.dropEvery = 0, this.corruptEvery = 0});

  final int dropEvery;
  final int corruptEvery;
  int _count = 0;

  Uint8List? apply(Uint8List frame) {
    _count++;
    if (dropEvery > 0 && _count % dropEvery == 0) return null;
    if (corruptEvery > 0 && _count % corruptEvery == 0) {
      final corrupted = Uint8List.fromList(frame);
      corrupted[corrupted.length - 1] ^= 1;
      return corrupted;
    }
    return frame;
  }
}

File: core-protocol/src/main/kotlin/in/meshsetu/protocol/ProtocolMetrics.kt
package `in`.meshsetu.protocol

import java.io.Writer

data class ProtocolMetric(val timeMs: Long, val eventId: String? = null, val peer: String? = null, val kind: String, val value: Long? = null, val detail: String? = null)

class JsonLineMetricSink(private val writer: Writer) {
    @Synchronized fun write(metric: ProtocolMetric) {
        fun quote(value: String) = "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n") + "\""
        val fields = mutableListOf("\"time_ms\":${metric.timeMs}", "\"kind\":${quote(metric.kind)}")
        metric.eventId?.let { fields += "\"event_id\":${quote(it)}" }
        metric.peer?.let { fields += "\"peer\":${quote(it.take(12))}" }
        metric.value?.let { fields += "\"value\":$it" }
        metric.detail?.let { fields += "\"detail\":${quote(it)}" }
        writer.append("{").append(fields.joinToString(",")).append("}\n").flush()
    }
}

class LossyFrameInterceptor(private val dropEvery: Int = 0, private val corruptEvery: Int = 0) {
    private var count = 0
    fun apply(frame: ByteArray): ByteArray? {
        count++
        if (dropEvery > 0 && count % dropEvery == 0) return null
        if (corruptEvery > 0 && count % corruptEvery == 0) return frame.copyOf().also { it[it.lastIndex] = (it.last().toInt() xor 1).toByte() }
        return frame
    }
}

